### DISCUSSION TITLE
Fetch error when building production locally.

### DISCUSSION THREAD

**Discussion by AlexTechNoir** [vercel/next.js#14505] [Help] [answered]
Greetings, everyone!

I've made a demo e-commerce project for portfolio. I'm trying to launch prod version locally. When I run 'npm run build' it gives fetch error:
```
D:\React\next_online_store>npm run build

> next_online_store@0.1.0 build D:\React\next_online_store
> next build

info  - Loaded env from D:\React\next_online_store\.env.local
Warning: Built-in CSS support is being disabled due to custom CSS configuration being detected.
See here for more info: https://err.sh/next.js/built-in-css-disabled

> Using external babel configuration
> Location: "D:\React\next_online_store\package.json"
Creating an optimized production build

Compiled successfully.

Automatically optimizing pages .
Error occurred prerendering page "/sales/Mobile%20Phones". Read more: https://err.sh/next.js/prerender-error
FetchError: request to http://localhost:3000/api/data failed, reason: connect ECONNREFUSED 127.0.0.1:3000

Error occurred prerendering page "/sales/Laptops". Read more: https://err.sh/next.js/prerender-error
FetchError: request to http://localhost:3000/api/data failed, reason: connect ECONNREFUSED 127.0.0.1:3000

Error occurred prerendering page "/sales/Tablets". Read more: https://err.sh/next.js/prerender-error
FetchError: request to http://localhost:3000/api/data failed, reason: connect ECONNREFUSED 127.0.0.1:3000

> Build error occurred
Error: Export encountered errors on following paths:
        /sales/Laptops
        /sales/Mobile%20Phones
        /sales/Tablets
    at exportApp (D:\React\next_online_store\node_modules\next\dist\export\index.js:22:1166)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async build (D:\React\next_online_store\node_modules\next\dist\build\index.js:36:218)
Automatically optimizing pages .npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! next_online_store@0.1.0 build: `next build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the next_online_store@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Noir\AppData\Roaming\npm-cache\_logs\2020-06-23T14_15_21_869Z-debug.log
```

The errors have occurred on 3 statically generated pages w/dynamic routes.

**pages/sales/[category].js:**
```import React from 'react'
import Head from 'next/head'
import styled from 'styled-components'

import Layout from '../../components/Layout'
import Timer from '../../components/Timer'
import ProductListItem from '../../components/ProductListItem'

export async function getStaticPaths() {
  return {
    paths: [
      { params : { category: 'Mobile Phones' } },
      { params : { category: 'Laptops' } },
      { params : { category: 'Tablets' } }
    ],
    fallback: false
  }
}

export async function getStaticProps({ params }) {
  const res = await fetch('http://localhost:3000/api/data')    //<-- Here
  const data = await res.json()

  return { props: { params, data } }
}

export default function Sale({ params, data }) {
  return (
    <React.Fragment>
      <Head>
        <title>{params.category} Sale! - Alimazon</title>
        <meta
          name="description"
          content={`${params.category} on SALE right now!!!`}
        />
      </Head>

      <Layout>
        <DivSales>
          <div>
            <img
              src={`/img/carousel/${params.category}/01.webp`}
              alt={`${params.category} Sale`}
            />
            <hr />
            <Timer />
          </div>
          <div>
            {
              data
                .filter(dataItem =>
                  dataItem.category === params.category && dataItem.hasDiscount === true
                )
                .map(dataItem => {
                  return <ProductListItem key={dataItem.id} dataItem={dataItem} />                  
                })
            }
          </div>
        </DivSales>
      </Layout>
    </React.Fragment>
  )
}

const DivSales = styled.div`
  grid-area: 2 / 2 / 3 / 3;
  justify-self: center;
  display: flex;
  flex-direction: column;    
  width: 100%;
  max-width: 1200px;
  > :first-child {
    display: flex;
    flex-direction: column;
    align-self: center;
    width: 100%;
    max-width: 960px;
    > img {
      width: 100%;
      height: 100%;
    }
    > :last-child {
      margin: auto;
      > :first-child {
        display: inline-flex;
        flex-direction: column;
        align-items: center;
      }
    }
  }
  > :last-child {
    display: grid;
    grid-template-columns: repeat(auto-fit, 290px);
    grid-row-gap: .5rem;
    grid-column-gap: .5rem;
    justify-content: center;	
    margin-top: 1em;
  }
`
```

Data comes from **pages/api/data.js**:

```
import { data } from '../../data'
...
  } else {
    res.status(200).json(data)
  }
...
```

According to [Prerender Error description](https://github.com/vercel/next.js/blob/master/errors/prerender-error.md) it looks like the props are not available for the pages. But why? (Or maybe it's something else?)

I know that it's not a real life example and such pages most likely should be SSRed for big e-com websites, but I wanted to understand how SG works in Next.js and here it's assumed that pages with sales are not updated frequently.

In dev mode everything's okay.

**Comment by jamesmosier** [comment, marked as answer]
The reason it is failing is because `http://localhost:3000/api/data` is not online when you are running your build.

Instead of trying to call an API route in `getStaticProps` you should just put your logic directly in the `getStaticProps` function. 

Or if you are just getting your data from a static file (i.e. `import { data } from '../../data'`) just import that in your `pages/sales/[category].js` page and then use it in `getStaticProps`. Next will eliminate that code from any client side bundles as long as you only use it within `getStaticProps`.


```js
// ...all of your other imports...
import { data } from '../../data'

export async function getStaticProps({ params }) {
  return { props: { params, data } }
}

// ...your other code...
```

**Reply by AlexTechNoir** [reply]
Thank you, James! Now I fully understand.

The docs said that I shouldn't fetch an API route in getStaticProps, but I thought it'll be okay if fetched data won't be changing. Turns out it's not the case at all.

About importing data directly - yeah, I kinda knew that from the beginning, but since I know little about backend so far, I wanted to emulate it as closely as possible to understand the basics of how frontend works with backend :)

**Reply by kunwar-11** [reply]
Hey can you please explain a bit 🙏

**Reply by AlexTechNoir** [reply]
Hi, Rishabh, sorry for long answer!

Next.js docs say that you should not call Next.js API route in getStaticProps. That's because getStaticProps are running during build when Next.js API routes are not available yet. Hence the error. But fetching from an external API is OK. You can put server code directly in getStaticProps. [Details](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation).

**Reply by kunwar-11** [reply]
So we can call some external API right ?

**Reply by kunwar-11** [reply]
> Hi, Rishabh, sorry for long answer!
> 
> Next.js docs say that you should not call Next.js API route in getStaticProps. That's because getStaticProps are running during build when Next.js API routes are not available yet. Hence the error. But fetching from an external API is OK. You can put server code directly in getStaticProps. [Details](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation).
> 

And this thing remains same for both serversideprops and staicprops ?

**Reply by AlexTechNoir** [reply]
The answer for both questions is "yes". The docs I linked say so.

**Reply by rhadu** [reply]
a possible "fix" that i discovered when i encountered this problem was that i ran "yarn serve" on a different terminal and "yarn build" after. This way `http://localhost:3000/api/` is reachable.

**Reply by DenisLopatin** [reply]
I've been looking for information about this for too long. I used to think that the api folder could be used as a wrapper over a third-party api, but apparently this is not the case if `getStaticProps` is used

**Reply by roman-flz** [reply]
Any possibility for implementing **on-demand revalidation** when querying MongoDB data from within getStaticProps!?

**Reply by renchris** [reply]
how do we do this in app router?

**Reply by kmgdevelopment** [reply]
> Instead of trying to call an API route in getStaticProps you should just put your logic directly in the getStaticProps function.

Can you expand on this? I'm pulling in data from a headless CMS running on a separate local server, I'm not using static data imports. Is there a way to run the NextJS build process with a successful fetch in this instance? I've tried the `npm run dev` then `npm run build` in a separate terminal window trick without success.

**Comment by frankxii** [comment]
I aslo met this problem and solved it by moving request statement from getStaticProps() to getServerSideProps().

follow the [offical document](https://nextjs.org/docs/basic-features/data-fetching), getStaticProps fetch data at build time and getServerSideProps fetch data on each request, so if we would like to fetch data when ervery time user visit page , we should use getServerSideProps() function.

**Reply by BlakeBrown** [reply]
I'm having this issue not in `getStaticProps` or `getServerSideProps`, but when calling one API route from another API route. Has anyone else encountered this? Next.js version 11

**Comment by gr007-40** [comment]
I was getting the same error with `getServerSideProps` and it resolved when I used `127.0.0.1` as a replacement for `localhost`.

**Reply by PritiKachiya** [reply]
Thanks a lot @Parzival-40  It's work for me 
i am also getting this error repeatedly , it's very useful for me

**Reply by Mezeobidi** [reply]
Please, I'm still having the issue. I urgently and desperately need help to deploy to cloudflare pages

import Head from 'next/head';
import { makeApiRequest } from '../../helpers/functionals';


export async function getServerSideProps(context) {
    const { query } = context;
    const studentRoute = `/v1/students/${query.id}`;
    return await makeApiRequest(context, studentRoute);
}

export const config = {
  runtime: 'edge',
};

export default function Testing({ responseData }) {
    const { data } = responseData;
    return (
        <div>
            <Head>
                <title>Student |  Details</title>
            </Head>

            <main className=py-1>
                
            </main>
        </div>
    )
}


I'm getting this error
ReferenceError: self is not defined
 Failed to collect page data for /student_details/[id]

I'm on nextjs version 12.3
